I managed to reproduce the problem with a small self-contained script:

```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import tensorflow as tf
import tensorflow.contrib.eager as tfe

config = tf.ConfigProto()
config.gpu_options.allow_growth = True

tf.enable_eager_execution(config=config)

shape = (5, 5)
initializer = tf.random_uniform_initializer(-1.0, 1.0)
var = tfe.Variable(initializer(shape), name='tensor')

with tf.GradientTape() as tape:
        a = tf.gather(tf.gather(var, [0, 1]), [0, 1])
        b = tf.gather(tf.gather(var, [2, 3]), [0, 1])

        x = tf.reshape(tf.concat((a[0], a[1]), axis=0), [1, -1])
        y = tf.reshape(tf.concat((b[0], b[1]), axis=0), [1, -1])

        r = tf.einsum('ij,ij->i', x, y)

g = tape.gradient(r, [var])
```

Here's the output - I'll commit a unit test shortly.

```bash
$ ./fun.py 
2019-04-19 20:27:55.511171: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2019-04-19 20:27:55.580302: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:964] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2019-04-19 20:27:55.580656: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1432] Found device 0 with properties: 
name: TITAN X (Pascal) major: 6 minor: 1 memoryClockRate(GHz): 1.531
pciBusID: 0000:02:00.0
totalMemory: 11.91GiB freeMemory: 5.75GiB
2019-04-19 20:27:55.580679: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1511] Adding visible gpu devices: 0
2019-04-19 20:27:55.921738: I tensorflow/core/common_runtime/gpu/gpu_device.cc:982] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-04-19 20:27:55.921768: I tensorflow/core/common_runtime/gpu/gpu_device.cc:988]      0 
2019-04-19 20:27:55.921779: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1001] 0:   N 
2019-04-19 20:27:55.921949: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1115] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 5526 MB memory) -> physical GPU (device: 0, name: TITAN X (Pascal), pci bus id: 0000:02:00.0, compute capability: 6.1)
Traceback (most recent call last):
  File "./fun.py", line 25, in <module>
    g = tape.gradient(r, [var])
  File "/home/anaconda/anaconda3/lib/python3.6/site-packages/tensorflow/python/eager/backprop.py", line 901, in gradient
    output_gradients=output_gradients)
  File "/home/anaconda/anaconda3/lib/python3.6/site-packages/tensorflow/python/eager/imperative_grad.py", line 64, in imperative_grad
    output_gradients)
  File "/home/anaconda/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 858, in grad_fun
    return [dresult._copy(device_name=self_device)]
AttributeError: 'IndexedSlices' object has no attribute '_copy'
```

--- a/tensorflow/python/framework/ops_test.py
+++ b/tensorflow/python/framework/ops_test.py
@@ -156,6 +156,19 @@ class IndexedSlicesTest(test_util.TensorFlowTestCase):
     tensor = ops.convert_to_tensor(x, name="tensor")
     self.assertAllEqual(self.evaluate(tensor), [[2, 3], [0, 0], [5, 7]])
 
+  @test_util.run_in_graph_and_eager_modes
+  def testEagerCopy(self):
+    shape = (5, 5)
+    initializer = tf.random_uniform_initializer(-1.0, 1.0)
+    var = tfe.Variable(initializer(shape), name='tensor')
+    with tf.GradientTape() as tape:
+        a = tf.gather(tf.gather(var, [0, 1]), [0, 1])
+        b = tf.gather(tf.gather(var, [2, 3]), [0, 1])
+        r = tf.einsum('ij,ij->i', a, b)
+    g = tape.gradient(r, [var])[0]
+    values = g.values if isinstance(g, tf.IndexedSlices) else g
+    self.assertAllEqual(values.get_shape(), (4, 5))
+
   @test_util.run_deprecated_v1
   def testNegation(self):
     with self.cached_session():

--- a/tensorflow/python/framework/ops_test.py
+++ b/tensorflow/python/framework/ops_test.py
@@ -156,18 +156,18 @@ class IndexedSlicesTest(test_util.TensorFlowTestCase):
     tensor = ops.convert_to_tensor(x, name="tensor")
     self.assertAllEqual(self.evaluate(tensor), [[2, 3], [0, 0], [5, 7]])
 
-  @test_util.run_in_graph_and_eager_modes
   def testEagerCopy(self):
-    shape = (5, 5)
-    initializer = tf.random_uniform_initializer(-1.0, 1.0)
-    var = tfe.Variable(initializer(shape), name='tensor')
-    with tf.GradientTape() as tape:
+    with context.eager_mode():
+      shape = (5, 5)
+      initializer = tf.random_uniform_initializer(-1.0, 1.0)
+      var = tfe.Variable(initializer(shape), name='tensor')
+      with tf.GradientTape() as tape:
         a = tf.gather(tf.gather(var, [0, 1]), [0, 1])
         b = tf.gather(tf.gather(var, [2, 3]), [0, 1])
         r = tf.einsum('ij,ij->i', a, b)
-    g = tape.gradient(r, [var])[0]
-    values = g.values if isinstance(g, tf.IndexedSlices) else g
-    self.assertAllEqual(values.get_shape(), (4, 5))
+      g = tape.gradient(r, [var])[0]
+      values = g.values if isinstance(g, tf.IndexedSlices) else g
+      self.assertAllEqual(values.get_shape(), [4, 5])
 
   @test_util.run_deprecated_v1
   def testNegation(self):

--- a/tensorflow/python/framework/ops_test.py
+++ b/tensorflow/python/framework/ops_test.py
@@ -26,6 +26,7 @@ import weakref
 from tensorflow.core.framework import attr_value_pb2
 from tensorflow.core.protobuf import config_pb2
 from tensorflow.python.client import session
+from tensorflow.python.eager import backprop
 from tensorflow.python.eager import context
 from tensorflow.python.eager import function as eager_function
 from tensorflow.python.framework import common_shapes
@@ -47,6 +48,7 @@ from tensorflow.python.ops import control_flow_ops
 from tensorflow.python.ops import math_ops
 from tensorflow.python.ops import resource_variable_ops
 from tensorflow.python.ops import resources
+from tensorflow.python.ops import special_math_ops
 from tensorflow.python.ops import variable_scope
 from tensorflow.python.ops import variables
 import tensorflow.python.ops.gradients  # pylint: disable=unused-import
@@ -158,16 +160,14 @@ class IndexedSlicesTest(test_util.TensorFlowTestCase):
 
   def testEagerCopy(self):
     with context.eager_mode():
-      shape = (5, 5)
-      initializer = tf.random_uniform_initializer(-1.0, 1.0)
-      var = tfe.Variable(initializer(shape), name='tensor')
-      with tf.GradientTape() as tape:
-        a = tf.gather(tf.gather(var, [0, 1]), [0, 1])
-        b = tf.gather(tf.gather(var, [2, 3]), [0, 1])
-        r = tf.einsum('ij,ij->i', a, b)
+      var = variables.Variable([[0.0], [0.0], [0.0], [0.0], [0.0]], name='tensor')
+      with backprop.GradientTape() as tape:
+        a = array_ops.gather(array_ops.gather(var, [0, 1]), [0, 1])
+        b = array_ops.gather(array_ops.gather(var, [2, 3]), [0, 1])
+        r = special_math_ops.einsum('ij,ij->i', a, b)
       g = tape.gradient(r, [var])[0]
-      values = g.values if isinstance(g, tf.IndexedSlices) else g
-      self.assertAllEqual(values.get_shape(), [4, 5])
+      values = g.values if isinstance(g, ops.IndexedSlices) else g
+      self.assertAllEqual(values.get_shape(), [4, 1])
 
   @test_util.run_deprecated_v1
   def testNegation(self):

